compiler/rustc_data_structures/src/sync/lock.rs

```diff
@@ -3,224 +3,257 @@
-                if self.opt.cell.get() {
-                    false
-                } else {
-                    self.opt.cell.set(true);
-                    true
+            match assume {
+                Assume::NoSync => {
+                    let LockMode::NoSync(cell) = &self.mode else {
+                        unsafe { unreachable_unchecked() }
+                    };
+                    if let Some(v) = no_sync(cell) {
+                        v
+                    } else {
+                        // Call this here instead of in `no_sync` so `track_caller` gets properly
+                        // passed along.
+                        lock_held()
+                    }
+                }
+                Assume::Sync => {
+                    let LockMode::Sync(lock) = &self.mode else {
+                        unsafe { unreachable_unchecked() }
+                    };
+                    sync(lock)
                 }
-            } else {
-                self.opt.lock.try_lock()
             }
         }
-    }
 
-    #[inline(always)]
-    fn lock(&self) {
-        if super::ERROR_CHECKING {
-            // We're in the debugging mode, so assert that the lock is not held so we
-            // get a panic instead of waiting for the lock.
-            assert_eq!(self.try_lock(), true, "lock must not be hold");
-        } else {
-            // SAFETY: This is safe since the union fields are used in accordance with `self.sync`.
+        #[inline(always)]
+        pub fn try_lock(&self) -> Option<LockGuard<'_, T>> {
+            let assume = self.mode.to_assume();
             unsafe {
-                if likely(!self.sync) {
-                    if unlikely(self.opt.cell.replace(true)) {
-                        cold_path(|| panic!("lock was already held"))
-                    }
-                } else {
-                    self.opt.lock.lock();
-                }
+                self.dispatch(
+                    assume,
+                    |cell| Some((cell.get() != LOCKED).then(|| cell.set(LOCKED)).is_some()),
+                    RawMutex::try_lock,
+                )
+                .then(|| LockGuard { lock: self, marker: PhantomData, assume })
             }
         }
-    }
 
-    /// This unlocks the lock.
-    ///
-    /// Safety
-    /// This method may only be called if the lock is currently held.
-    #[inline(always)]
-    unsafe fn unlock(&self) {
-        // SAFETY: The union use is safe since the union fields are used in accordance with
-        // `self.sync` and the `unlock` method precondition is upheld by the caller.
-        unsafe {
-            if likely(!self.sync) {
-                debug_assert_eq!(self.opt.cell.get(), true);
-                self.opt.cell.set(false);
-            } else {
-                self.opt.lock.unlock();
+        #[inline(always)]
+        #[track_caller]
+        pub unsafe fn lock_assume(&self, assume: Assume) -> LockGuard<'_, T> {
```

Comment:
I think this unsafe func needs a comment, too. Like `Assume` should consistent wiht `self.mode`